File: app/src/main/java/com/example/octodroid/views/components/LinearLayoutLoadMoreListener.java
package com.example.octodroid.views.components;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

public abstract class LinearLayoutLoadMoreListener extends RecyclerView.OnScrollListener {

    private int prevTotalCount = 0;

    private boolean isLoading = true;

    private LinearLayoutManager layoutManager;

    public LinearLayoutLoadMoreListener(LinearLayoutManager linearLayoutManager) {
        this.layoutManager = linearLayoutManager;
    }

    @Override
    public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
        super.onScrolled(recyclerView, dx, dy);

        int visibleItemCount = recyclerView.getChildCount();
        int totalItemCount = layoutManager.getItemCount();
        int firstVisibleItem = layoutManager.findFirstVisibleItemPosition();

        if (isLoading) {
            if (prevTotalCount < totalItemCount) {
                isLoading = false;
                prevTotalCount = totalItemCount;
            }
        } else {
            if (firstVisibleItem + visibleItemCount >= totalItemCount) {
                onLoadMore();
                isLoading = true;
            }
        }
    }

    public abstract void onLoadMore();
}

File: app/src/main/java/com/example/octodroid/views/components/LinearLayoutLoadMoreListener.kt
package com.example.octodroid.views.components

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView

abstract class LinearLayoutLoadMoreListener(private val layoutManager: LinearLayoutManager) : RecyclerView.OnScrollListener() {

    private var prevTotalCount = 0

    private var isLoading = true

    override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
        super.onScrolled(recyclerView, dx, dy)

        val visibleItemCount = recyclerView!!.getChildCount()
        val totalItemCount = layoutManager.getItemCount()
        val firstVisibleItem = layoutManager.findFirstVisibleItemPosition()

        if (isLoading) {
            if (prevTotalCount < totalItemCount) {
                isLoading = false
                prevTotalCount = totalItemCount
            }
        } else {
            if (firstVisibleItem + visibleItemCount >= totalItemCount) {
                onLoadMore()
                isLoading = true
            }
        }
    }

    abstract fun onLoadMore()
}

